src/chainstate/stacks/boot/pox.clar

```diff
@@ -0,0 +1,754 @@
+(define-constant MIN-POX-REWARD-CYCLES u1)
+(define-constant MAX-POX-REWARD-CYCLES u12)
+
+;; Default length of the PoX registration window, in burnchain blocks.
+(define-constant REGISTRATION-WINDOW-LENGTH u250)
+
+;; Default length of the PoX reward cycle, in burnchain blocks.
+(define-constant REWARD-CYCLE-LENGTH u1000)
+
+;; Valid values for burnchain address versions.
+;; These correspond to address hash modes in Stacks 2.0.
+(define-constant ADDRESS-VERSION-P2PKH u0)
+(define-constant ADDRESS-VERSION-P2SH u1)
+(define-constant ADDRESS-VERSION-P2WPKH u2)
+(define-constant ADDRESS-VERSION-P2WSH u3)
+
+;; Stacking thresholds
+(define-constant STACKING-THRESHOLD-25 u20000)
+(define-constant STACKING-THRESHOLD-100 u5000)
+
+;; PoX disabling threshold (a percent)
+(define-constant POX-REJECTION-FRACTION u25)
+
+;; Data vars that store a copy of the burnchain configuration.
+;; Implemented as data-vars, so that different configurations can be
+;; used in e.g. test harnesses.
+(define-data-var pox-registration-window-length uint REGISTRATION-WINDOW-LENGTH)
+(define-data-var pox-reward-cycle-length uint REWARD-CYCLE-LENGTH)
+(define-data-var pox-rejection-fraction uint POX-REJECTION-FRACTION)
+(define-data-var first-burnchain-block-height uint u0)
+(define-data-var configured bool false)
+
+;; This function can only be called once, when it boots up
+(define-public (set-burnchain-parameters (first-burn-height uint) (pox-reg-window-len uint) (pox-reward-cycle-len uint) (pox-rejection-frac uint))
+    (if (and is-in-regtest (not (var-get configured)))
+        (begin
+            (var-set first-burnchain-block-height first-burn-height)
+            (var-set pox-registration-window-length pox-reg-window-len)
+            (var-set pox-reward-cycle-length pox-reward-cycle-len)
+            (var-set pox-rejection-fraction pox-rejection-frac)
+            (var-set configured true)
+            (ok true)
+        )
+        (ok false)
+    )
+)
+
+;; The Stacking lock-up state and associated metadata.
+;; Records can be inserted into this map via one of two ways:
+;; * via contract-call? to the (stack-stx) method, or
+;; * via a transaction in the underlying burnchain that encodes the same data.
+;; In the latter case, this map will be updated by the Stacks
+;; node itself, and transactions in the burnchain will take priority
+;; over transactions in the Stacks chain when processing this block.
+(define-map stacking-state
+    ((stacker principal))
+    (
+        ;; how many uSTX locked?
+        (amount-ustx uint)
+        ;; Description of the underlying burnchain address that will
+        ;; receive PoX'ed tokens. Translating this into an address
+        ;; depends on the burnchain being used.  When Bitcoin is
+        ;; the burnchain, this gets translated into a p2pkh, p2sh,
+        ;; p2wpkh-p2sh, or p2wsh-p2sh UTXO, depending on the version.
+        (pox-addr (tuple (version uint) (hashbytes (buff 20))))
```

Comment:
I think `version` should be a `(buff 1)` rather than a uint.

Reply:
Sure thing.

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1850/commits/9e3899c47e950996c32d0bece74df992c5768d91